pacman_contest_cluster.py

```diff
@@ -23,307 +23,754 @@
+        with open(args.build_config_file, "w") as f:
+            json.dump(settings, f, sort_keys=True, indent=4, separators=(",", ": "))
 
     return settings
 
 
 # ----------------------------------------------------------------------------------------------------------------------
-if __name__ == '__main__':
-    settings = load_settings()
 
 
-    # from getpass import getuser
+def list_partition(list_in, n):
+    # partitions a list into n (nearly) equal lists: https://stackoverflow.com/questions/3352737/how-to-randomly-partition-a-list-into-n-nearly-equal-parts
+    random.shuffle(list_in)
+    return [list_in[i::n] for i in range(n)]
+
+
+def get_agent_factory(team_name):
+    """returns the agent factory for a given team"""
+    return os.path.join(TEAMS_SUBDIR, team_name, AGENT_FACTORY)
+
+
+# @dataclass
+# class ContestSettings:
+#     organizer: str
+#     teams_root: str
+#     staff_teams_vs_others_only: bool
+#     include_staff_team: bool
+#     staff_teams_dir
+#     compress_logs
+#     max_steps
+#     no_fixed_layouts,
+#     fixed_layouts_file
+#     no_random_layouts
+#     team_names_file
+#     allow_non_registered_students
+#     ignore_file_name_format
+#     www_dir
+#     fixed_layout_seeds=[]
+#     random_seeds=[]
+#     stats_archive_dir=None
+#     logs_archive_dir=None
+#     replays_archive_dir=None
+#     upload_replays=False
+#     upload_logs=False
+#     split=1
+
+
+class MultiContest:
+    def __init__(self, settings):
+        self.layouts = set()
+        self.split = settings["split"]
+        self.settings = settings
+
+        if not os.path.exists(os.path.join(DIR_SCRIPT, CONTEST_ZIP_FILE)):
+            logging.error(f"Contest zip file {CONTEST_ZIP_FILE} could not be found. Aborting.")
+            sys.exit(1)
 
-    # prompt for password (for password authentication or if private key is password protected)
-    # hosts = [Host(no_cpu=2, hostname='localhost', username=getuser(), password=getpass(), key_filename=None)]
-    # use this if no pass is necessary (for private key authentication)
-    # hosts = [Host(no_cpu=2, hostname='localhost', username=getuser(), password=None, key_filename=None)]
+        if not settings["fixed_layouts_file"]:
+            logging.error(f"Layouts file {settings['fixed_layouts_file']} could not be found. Aborting.")
+            sys.exit(1)
 
-    with open(settings['workers_file'], 'r') as f:
-        workers_details = json.load(f)['workers']
+        self.tmp_contest_dir = os.path.join(TMP_DIR, TMP_CONTEST_DIR)
+
+        # Setup Pacman CTF environment by extracting it from a clean zip file
+        self._prepare_platform(
+            os.path.join(DIR_SCRIPT, CONTEST_ZIP_FILE),
+            settings["fixed_layouts_file"],
+            self.tmp_contest_dir,
+            settings["no_fixed_layouts"],
+            settings["no_random_layouts"],
+            settings.get("fixed_layout_seeds", []),
```

Comment:
I changed this as these two seed options may not exist in a run that is not resuming

Reply:
This seems to be working but leave it here in case I did not get it and made a mistake for Andrew to double check.

Reply:
Yes, this should do the trick for now. In the medium term (i.e. after the final contest), I would really like to clean up settings a lot, using dataclasses for type safety but also to perform all the validation like this in one place. So you would know that when you have the settings object, all fields are initialised properly (including to defaults), no fields are inconsistent (i.e. giving fixed seeds than there are layouts) etc etc. I've made #79 to track this

Reply:
Just perfect; I see the new issue you created, good!